### PR TITLE
feat: HMR support for serverMiddleware

### LINKED ISSUE #6877: Don't rebuild full Nuxt project on serverMiddleware changes
### What problem does this feature solve?

Currently Nuxt rebuilds the full project when you touch any serverMiddleware. Because this reloads the full Nuxt server, any file listeners are removed during the build causing Nuxt to not detect file changes during the build process (which can often take 10-30s).

### What does the proposed changes look like?

Dont rebuild the webpack bundle if only serverMiddleware is touched

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c10207">#c10207</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by pi0**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Refactor

## Description

Fixes #6877 #1509.

This PR brings huge DX improvements to Nuxt [serverMiddleware](https://nuxtjs.org/api/configuration-servermiddleware) by watching all serverMiddleware (and their require dependency tree). HMR supported for both path and handler.

Changes summary:
- Server refactored to support in-place HMR of middleware stack and also adds additional meta (7505ac5b7a7d9e16ec0e30023109cefbb8bc50f9)
  - New middleware syntax is more consistent with express (path/handler ~> route/handle - with backward compatibility)
- Fixed an important bug in cjs helper that causes process fatal error when accessing a cache entry which had already errors (more DX coming in another PR for HMR of nuxt.config) (26581e65b9ffa9cbbe84a9e59d8d8bf83fa7fb9b)
- Replaced old full restart in Builder with new HMR in Builder (In rare cases if any error happens fullRestart fallback happens) (8a950ce990a7cfd737a4e5592abce4e327f2a637)

**Error handling:** Normal runtime errors already being handled by error middleware. In case an error happens due to syntax/reference error, a placeholder will be automatically replaced. If path is defined inside same file, we try to preserve last state otherwise show error on all routes until problem is fixes.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/builder/src/builder.js [hunk 1]
```diff
@@ -690,25 +695,61 @@ export default class Builder {
...
-        if (isString(serverMiddleware)) {
-          return serverMiddleware
+  serverMiddlewareHMR () {
+    // Check nuxt.server dependency
+    if (!this.nuxt.server) {
+      return
+    }
+
+    // Get registered server middleware with path
+    const entires = this.nuxt.server.serverMiddlewarePaths()
```
Comment (Atinux) [inline packages/builder/src/builder.js]:
```suggestion
    const entries = this.nuxt.server.serverMiddlewarePaths()
```

File: packages/builder/src/builder.js [hunk 2]
```diff
@@ -690,25 +695,61 @@ export default class Builder {
...
+
+    // Resolve dependency tree
+    const deps = new Set()
+    const dep2Entry = {}
+
+    for (const entry of entires) {
```
Comment (Atinux) [inline packages/builder/src/builder.js]:
```suggestion
    for (const entry of entries) {
```

**Review by atinux — CHANGES_REQUESTED** [review]
Small typo

**Review by atinux — APPROVED** [review]

**Comment by Atinux** [comment]
Can we add maybe a warn for path usage? Or update the docs for serverMiddleware?

**Comment by pi0** [comment]
@Atinux Internally we normalize path/handler without any warning by this PR to use connect/express convention. We can deprecate it for Nuxt3.

**Comment by GaborTorma** [comment]
Please update the docs woth examples, because it not working for me.
Thanx!

**Comment by jgmullor** [comment]
I have the same issue as @GaborTorma, only the first change on the server middlewares is detected

**Comment by pi0** [comment]
@GaborTorma @jgmullor 

- If you are using Windows, there is a known issue going to be fixed with next release.
- This is a basic DX improvement for serverMiddleware. Nuxt functions coming soon ;)
- Would you please create an issue with reproduction steps when it is not correctly working?

**Comment by jgmullor** [comment]
@pi0 in our case we're using OSX/Ubuntu, I will try to create an issue with a jsbin later today, thank you! =)

**Comment by HuddleHouse** [comment]
Does anybody have an example of this working? 

I am getting this error: 

```
this.nuxt.server.serverMiddlewarePaths is not a function
```

**Comment by pi0** [comment]
@HuddleHouse Would you please ensure having 2.12+ version of `@nuxt/server` installed in `node_modules`? Are you using nuxt as middleware?

**Comment by HuddleHouse** [comment]
@pi0 Thanks for responding! That helped me figure out what was wrong. 

It turns out 2.11+ versions of `@nuxt/server` & `@nuxt/core` were getting installed. I installed 2.12+ versions in dev dependencies and it worked!

**Comment by humanistical** [comment]
I'm getting the below error after this change, using an Express api backend exported as below:
~/api/index.js:
import express from 'express'
...
const app = express()
...
module.exports = {
  route: '/api',
  handler: app
}

nuxt.config.js:
serverMiddleware: [{ route: '/api', handler: '~/api/index.js' }]

Error:
FATAL Cannot create property '_middleware' on string '~/api/index.js' 

at Server.resolveMiddleware (node_modules/@nuxt/server/dist/server.js:816:35) 
at Server.useMiddleware (node_modules/@nuxt/server/dist/server.js:822:36) 
at Server.setupMiddleware (node_modules/@nuxt/server/dist/server.js:680:12) 
at async Server.ready (node_modules/@nuxt/server/dist/server.js:610:5) 
at async Nuxt._init (node_modules/@nuxt/core/dist/core.js:689:7)

**Comment by mobihack** [comment]
@humanistical I am encountering the same problem. Any fix?

**Comment by TeleMediaCC** [comment]
> I'm getting the below error after this change, using an Express api backend exported as below:
> ~/api/index.js:
> import express from 'express'
> ...
> const app = express()
> ...
> module.exports = {
> route: '/api',
> handler: app
> }
> 
> nuxt.config.js:
> serverMiddleware: [{ route: '/api', handler: '~/api/index.js' }]
> 
> Error:
> FATAL Cannot create property '_middleware' on string '~/api/index.js'
> 
> at Server.resolveMiddleware (node_modules/@nuxt/server/dist/server.js:816:35)
> at Server.useMiddleware (node_modules/@nuxt/server/dist/server.js:822:36)
> at Server.setupMiddleware (node_modules/@nuxt/server/dist/server.js:680:12)
> at async Server.ready (node_modules/@nuxt/server/dist/server.js:610:5)
> at async Nuxt._init (node_modules/@nuxt/core/dist/core.js:689:7)

@mobihack 


You need to exports only the app in ~/api/index.js file:

`module.exports = app`

**Comment by mobihack** [comment]
@TeleMediaCC  Thank you. This solved my problem.

For others who may encounter this problem, you don't need to sub route to '/api' inside your express setup. Just export without prepending api to your express routes.

**Pull request merged**